server/src/main/java/com/cloud/acl/DomainChecker.java

```diff
@@ -167,81 +182,155 @@ public boolean checkAccess(User user, ControlledEntity entity) throws Permission
     }
 
     @Override
-    public boolean checkAccess(Account account, DiskOffering dof) throws PermissionDeniedException {
-        if (account == null || dof == null || dof.getDomainId() == null) {//public offering
-            return true;
+    public boolean checkAccess(Account account, DiskOffering dof, DataCenter zone) throws PermissionDeniedException {
+        boolean isAccess = false;
+        // Check fo domains
+        if (account == null || dof == null) {
+            isAccess = true;
         } else {
             //admin has all permissions
             if (_accountService.isRootAdmin(account.getId())) {
-                return true;
+                isAccess = true;
```

Comment:
why not return here?

Reply:
Because later in this method there is checkAccess for the zone as well. It might be possible that offering is not specified for the passed zone.